compiler/rustc_privacy/src/lib.rs

```diff
@@ -242,6 +242,12 @@ where
                 // This will also visit substs if necessary, so we don't need to recurse.
                 return self.visit_projection_ty(proj);
             }
+            ty::Alias(ty::Weak, ty::AliasTy { def_id, substs, .. }, ..) => {
+                // Skip repeated type aliases to avoid infinite recursion.
+                if self.visited_opaque_tys.insert(def_id) {
+                    return tcx.type_of(def_id).subst(tcx, substs).super_visit_with(self);
+                }
```

Comment:
I considered that. This makes them work like they did before. The only way to get a different type depending on the substs is via projections inside, and these are handled by the recursion.

I guess since we don't need to preserve the current behaviour, I should revisit the entire PR and make them less weak.